compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -237,12 +238,23 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                     fcx,
                 );
 
-                err.span_suggestion_verbose(
-                    self.expr_span.shrink_to_lo(),
-                    "dereference the expression",
-                    "*",
-                    Applicability::MachineApplicable,
-                );
+                if matches!(self.expr.kind, ExprKind::AddrOf(..)) {
+                    // get just the borrow part of the expression
+                    let span = self.expr_span.with_hi(self.expr.peel_borrows().span.lo());
```

Comment:
Wouldn't this be a problem if you had `&&&&&0u8 as &&&&u8`?

Reply:
That code compiles on stable, but the case: `&&&0u8 as &&u32` (note the different int type) fails to compile with a more generic error (without any suggestions). Ideally, it would understand that multiple borrows/derefs can be used but that is outside the scope of this PR and requires some thought.